### ISSUE TITLE
Working with multiple bootstrappers

### ISSUE THREAD

**Issue by sanurss** [Caliburn-Micro/Caliburn.Micro#297]
I have an system consisting of toolbar and multiple tools. Every tool is a separate application provided as a set of dll files. During startup system seeks for libraries containing bootstrappers and registers then as icons on toolbar. These applications can be delivered independently and can be run simultaneously.

After upgrade from Caliburn.Micro 1.5.2 to 2.0.2 a problem appeared. When I start application for the first time everything goes fine. However when I start it again from the toolbar it crashes on load with an exception.
```
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at Caliburn.Micro.AssemblySourceCache.<Install>b__4(Type t)
   at Caliburn.Micro.EnumerableExtensions.Apply[T](IEnumerable`1 enumerable, Action`1 action)
   at Caliburn.Micro.AssemblySourceCache.<Install>b__0(Object s, NotifyCollectionChangedEventArgs e)
   at System.Collections.ObjectModel.ObservableCollection`1.OnCollectionChanged(NotifyCollectionChangedEventArgs e)
   at Caliburn.Micro.BindableCollection`1.OnCollectionChanged(NotifyCollectionChangedEventArgs e)
   at Caliburn.Micro.BindableCollection`1.<>c__DisplayClasse.<AddRange>b__d()
   at Caliburn.Micro.XamlPlatformProvider.OnUIThread(Action action)
   at Caliburn.Micro.BindableCollection`1.AddRange(IEnumerable`1 items)
   at Caliburn.Micro.BootstrapperBase.StartRuntime()
   at Caliburn.Micro.BootstrapperBase.Initialize()
   at DemoApp.Starter.DemoUIStarter.Start()
   at ...
```

What is the way to solve this issue? Is there any solution to run applications with multiple bootstrappers safely?

**Comment by nigel-sampson** [comment]
I'm assuming you're working on WPF as a platform?

By bootstrapper are you referring to a class inheriting `BootstrapperBase`? So I'm assuming as part of this registration process you're calling `Initialize` on each bootstrapper?

Each bootstrapper is working to initialize the `AssemblySource` and the `AssemblySourceCache`.

``` csharp
var baseExtractTypes = AssemblySourceCache.ExtractTypes;

AssemblySourceCache.ExtractTypes = assembly =>
{
    var baseTypes = baseExtractTypes(assembly);
    var elementTypes = assembly.GetExportedTypes()
        .Where(t => typeof(UIElement).IsAssignableFrom(t));

    return baseTypes.Union(elementTypes);
};

AssemblySource.Instance.Refresh();
```

This code wouldn't cause errors if it was run multiple times, the trouble is that each time is layers more functionality on top of itself, the LINQ union operation should clean that up though since it remove the duplicates. Not the best though.

Most like the problem is in `StartRuntime` where there's the following.

``` csharp
AssemblySourceCache.Install();
AssemblySource.Instance.AddRange(SelectAssemblies());
```

The assembly source cache uses the `AssemblySourceCache.ExtractTypes` method defined above to extract the views and view models from the assemblies listed in `SelectAssemblies`. `AssemblySourceCache` looks for changes to `AssemblySource` and pre-extracts the types for later.

Most likely the problem is that multiple bootstrappers are returning the same assembly causing the type to be extracted multiple times.

Solution could be not returning anything from SelectAssemblies and having some other method of populating that cache.

**Comment by sanurss** [comment]
Yes, the app is based on WPF and I was referring to `BootstrapperBase`. Our problem is this permanent-static cache that cannot be cleared in any way. So next time we start the same tool/subapplication we end up with duplicates.

Our workaround at this moment is custom Caliburn.Micro build with `AssemblySourceCache.Install();` line commented and it seems to work quite fine. Anyway this is definitely a hack and unacceptable for production release.

There was no cache in Caliburn.Micro 1.5 and it worked great for us. It would be nice to have an option to disable cache. You've mentioned to populate the cache in a different way and today a new idea come to my mind that a change like this may help and it seems to work fine.
```
public Bootstrapper()
    : base(false)
{
    AssemblySource.Instance.Clear();
    Initialize();
}
```

Do you see any potential drawbacks of the code above?

**Comment by nigel-sampson** [comment]
At the moment there isn't an option to disable the cache?

Can I confirm in a number of these applications you're overrriding `SelectAssemblies`?

**Comment by sanurss** [comment]
I don't see such option in API, take a look at this file [AssemblySource.cs](/Caliburn-Micro/Caliburn.Micro/blob/master/src/Caliburn.Micro.Platform/AssemblySource.cs).

See my bootstrapper below (looks almost the same in all apps).
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using Caliburn.Micro;
using Castle.MicroKernel.Registration;
using Castle.Windsor;
using DemoApp.Gui.ViewModels;
using DemoApp.Gui.Views;
using DemoApp.Starter.Installers;

namespace DemoApp.Starter
{
    public class Bootstrapper : BootstrapperBase
    {
        private WindsorContainer _container;

        public Bootstrapper()
            : base(false)
        {
            Initialize();
        }

        protected override void Configure()
        {
            ViewLocator.AddNamespaceMapping("DemoApp.Gui.Views");

            _container = new WindsorContainer();
            _container.Install(new UiInstaller());
        }

        protected override object GetInstance(Type serviceType, string key)
        {
            if (string.IsNullOrEmpty(key))
            {
                return _container.Resolve(serviceType);
            }

            return _container.Resolve(key, serviceType);
        }

        protected override IEnumerable<Assembly> SelectAssemblies()
        {
            return new[] { typeof(MainView).Assembly };
        }

        protected override IEnumerable<object> GetAllInstances(Type serviceType)
        {
            return _container.ResolveAll(serviceType).Cast<object>();
        }

        public void Run()
        {
            DisplayRootViewFor<MainViewModel>();
        }
    }
}
```

**Comment by nigel-sampson** [comment]
Sorry I meant `SelectAssemblies` in the `Bootstrapper` which it appears you are. However you're returning what the default implementation would (assuming the bootstrapper is defined in the same assembly as `MainView`).

The error is most likely cause the second time the app is run and `AssemblySource.Instance.AddRange(SelectAssemblies());` is called passing the assembly in the cache for a second time.

I'd recommend overriding `SelectAssemblies` to only return things not already in the AssemblySource.

``` csharp
protected override IEnumerable<Assembly> SelectAssemblies()
{
    var viewAssembly = typeof (MainViewView).Assembly;

    if (!AssemblySource.Instance.Contains(viewAssembly))
        yield return viewAssembly;
}
```

**Comment by sanurss** [comment]
The change you proposed seems to work. However it is not very nice from maintenance point of view as it is spread out into multiple bootstrapper files. Anyway thanks for support :)

**Comment by aplocher** [comment]
I began getting this exception, but after reading this post I remembered that I commented out some old code (to see if it was still needed) but forgot about it, because I was only testing with one bootstrapper for the longest time.  Once I tried a second one (weeks later) I ran into this.  Uncommented my old code (below) and works great again lol.

    protected override IEnumerable<Assembly> SelectAssemblies()
    {
        //http://www.andyrace.com/2014/05/caliburn-micro-and-usercontrols.html
        return base.SelectAssemblies().Where(x => !AssemblySource.Instance.Contains(x));
    }

Thank you
